lsp-mode.el

```diff
@@ -3702,77 +3702,101 @@ and the position respectively."
+    (let* ((bounds (bounds-of-thing-at-point 'symbol))
+           (trigger-chars (->> (lsp--server-capabilities)
+                               (gethash "completionProvider")
+                               (gethash "triggerCharacters")))
+           start-point result done?)
+      (list
+       (if bounds (car bounds) (point))
+       (point)
+       (lambda (_probe pred action)
          (cond
-          (text-edit
-           (delete-region start-point (point))
-           (lsp--apply-text-edit text-edit))
-          (insert-text
-           (delete-region (- (point) (length candidate)) (point))
-           (insert insert-text)))
-
-         (when (eq insert-text-format 2)
-           (yas-expand-snippet (buffer-substring start-point (point))
-                               start-point
-                               (point)))
-         (when additional-text-edits
-           (lsp--apply-text-edits additional-text-edits)))
-
-       (when lsp-signature-auto-activate
-         (lsp-signature-activate))
-
-       (when (lsp--looking-back-trigger-characters-p trigger-chars)
-         (setq this-command 'self-insert-command))))))
+          ((eq action 'metadata)
+           `(metadata . ((display-sort-function . (lambda (candidates)
+                                                    (--sort (string-lessp (get-text-property 0 'lsp-sort-text it)
+                                                                          (get-text-property 0 'lsp-sort-text other))
+                                                            candidates))))))
+          ((eq (car-safe action) 'boundaries) nil)
+          (t (if done?
+                 result
+               (setf start-point (point))
+               (-let* ((resp (lsp-request-while-no-input "textDocument/completion"
+                                                         (plist-put (lsp--text-document-position-params)
+                                                                    :context (ht ("triggerKind" 1)))))
+                       (items (lsp--sort-completions (cond
+                                                      ((seqp resp) resp)
+                                                      ((hash-table-p resp) (gethash "items" resp)))))
+                       ((start . end) (or
+                                       (-when-let* ((item (lsp-elt items 0))
+                                                    ((&hash "textEdit" text-edit) item)
+                                                    ((&hash "range") text-edit))
+                                         (lsp--range-to-region range))
+                                       bounds))
+                       (probe
+                        (if (and start end) (buffer-substring-no-properties start end)
+                          "")))
+                 (setf done? (or (seqp resp)
+                                 (not (gethash "isIncomplete" resp)))
+                       result (--> (seq-into (--map (-let [(&hash "label"
```

Comment:
possible to move the inner --map in the --> and reduce the nesting here (e. g. start with `(--> items ...`?

Reply:
Done